src/main/kotlin/de/uniwuerzburg/omod/core/GridFactory.kt

```diff
@@ -174,4 +183,187 @@ fun makeClusterGrid(focusAreaPrecision: Double, buildings: List<Building>,
+    }
+
+    for ((nClusterDivisor, buildingsAtDistance) in buildingGroups) {
+        cells.addAll(
+            cluster(
+                focusAreaPrecision*nClusterDivisor, buildingsAtDistance, geometryFactory, transformer, startID)
+        )
+    }
+
+    return cells
+}
+
+/**
+ * Group buildings into traffic assignment zones (cells) with predefined geometry.
+ * Calculates cell centroids using building centroids
+ * Cells outside Focus area are not considered yet...
+ *
+ * @param tazFile geojson file with geometries of traffic assignment zones (cells)
+ * @param buildings the buildings which are assigned to cells
+ * @param transformer CRS transformer to use
+ *
+ * @return routing cells
+ */
+@OptIn(ExperimentalSerializationApi::class)
+@Suppress("unused")
+fun makeClusterGridFromFile(tazFile: File, buildings: List<Building>, transformer: CRSTransformer ) : List<Cell> {
+
+    // Get TAZs from file
+    val taz : GeoJsonNoProperties = json.decodeFromString(tazFile.readText(Charsets.UTF_8))
+
+    // extract shape of each zone
+    var zones = arrayOf<Geometry>()
+    if (taz is GeoJsonFeatureCollectionNoProperties) {
+        zones = taz.features.map { it.geometry.toJTS(geometryFactory) }.toTypedArray()
+    }
+
+    // Get Buildings in FocusArea
+    val focusAreaBuildings = buildings.filter { it.inFocusArea }
+
+    // Init empty cells
+    val grid = mutableListOf<Cell>()
+
+    // Init empty cell buildings arrays (one for each zone)
+    val cellBuildingLists = MutableList(zones.size) { mutableListOf<Building>() }
+
+    /*
+    val cellBuildingsTree = HPRtree()
+    for (building in focusAreaBuildings) {
+        cellBuildingsTree.insert(building.geometry.envelopeInternal, building)
+    }
+    */
+
+
+    // Structure for aggregated cell information
+    val simpleTAZs = mutableListOf<OutputSimpleTAZ>()
+
+    // Fill Cells with buildings
+
+    logger.info("Fill Focus Area Cells...")
+    for (b in focusAreaBuildings) {
+        zones.withIndex().forEach { (i, z) ->
+            if(z.contains(transformer.toLatLon( b.point ))) {
+                cellBuildingLists[i].add(b)
+            }
+        }
+    }
+
+    logger.info("Aggregate Cell Information to TAZs...")
+    var id = 0
+    for ((i,z) in zones.withIndex()) {
+
+        // Calc Cell Centroid from Building Areas
+        // if cell contains buildings
+        var centroid = arrayOf(0.0,0.0)
+        if(cellBuildingLists[i].isNotEmpty()) {
+            for (b in cellBuildingLists[i]) {
+                val point: DoubleArray = b.getPoint()
+                for (j in centroid.indices) {
+                    centroid[j] += point[j]
+                }
+            }
+            for (j in centroid.indices) {
+                centroid[j] /= cellBuildingLists[i].size.toDouble()
+            }
+        // if cell is empty
+        } else {
+            centroid = arrayOf(z.centroid.x, z.centroid.y)
+        }
+
+        val featureCentroid = Coordinate(centroid[0], centroid[1])
```

Comment:
You can calculate this a lot easier by creating a MultiPoint Object and getting the centroid from that:
val centroid = MultiPoint(points, factory).centroid()